internal/reflect/into.go

```diff
@@ -148,6 +148,14 @@ func BuildValue(ctx context.Context, typ attr.Type, val tftypes.Value, target re
 
 		return target, diags
 	}
+
+	// If the attr.Type has a dynamic value, then we need to use the tftypes.Value
+	// to determine the attr.Type for the reflection logic below.
+	dynTyp, ok := typ.(attr.TypeWithDynamicValue)
+	if ok {
+		typ = dynTyp.DetermineAttrType(val.Type())
```

Comment:
Is it worth leaving a comment mentioning that this is mutating the `typ attr.Type` parameter passed to `BuildValue()`?

Reply:
Good shout, added a comment to be more specific about the mutation